File: chameneos/src/chameneos.go
/* Compile and run under Linux and macOS with "go run chameneos.go"

Synchronous channels are used for passing the token between the node goroutines.
The mall and each chameneos is a goroutine. The mall has a single channel over
which it receives requests from chameneos. Each chameneos sends to the mall its
own color and a channel over which it can receive the color of the partner.

In Go, a program terminates when the main program terminates. Here, the main
program waits on the channel done for the mall goroutine to signal that it is
terminating.
*/

package main
import "fmt"

const Chams = 100
const Rounds = 1000

type Color int
const (blue Color = 0; red; yellow)

type Request struct{col Color; reply chan Color}

func chameneos(col Color, mall chan Request) {
  reply := make(chan Color)
  for {
    // in forest
    mall <- Request{col, reply}
    // waiting to meet
    otherCol := <- reply
    // changing color
    if col != otherCol {col = 3 - col - otherCol}
  }
}
func mall(cham chan Request, done chan bool){
  diff := 0
  for r := 0; r < Chams * Rounds / 2; r++ {
    fst := <- cham; snd := <- cham
    fst.reply <- snd.col; snd.reply <- fst.col
    if fst.col != snd.col {diff += 1}
  }
  fmt.Println("Color changes:", diff)
  done <- true
}
func main() {
  req := make(chan Request); done := make(chan bool)
  go mall(req, done)
  for i := 0; i < Chams; i++ {go chameneos(Color(i % 3), req)}
  <- done
}

File: chameneos/src/chameneos.rs
/* Compile under Linux and macOS with "rustc chameneos.rs"; run with
"./chameneos".

Synchronous channels from the mpsc library (multiple producer, single consumer)
are used for chameneos notifying the mall of their arrival and for the mall
sending back the color of the other chameneos. Rust forces that each chameneos
creates a new reply channel in each iteration. Alternatively, the reply channel
can be created once and then cloned, which can improve efficiency.

Rust requires all threads to terminate properly for the program to terminate
properly. When the mall is done pairing chameneos, it replies to each further
request by sending None instead of a color, which will cause the requesting
chameneos to terminate.
*/

use std::sync::mpsc::*;
use std::thread::spawn;

const CHAMS: u32 = 100;
const MEETS: u32 = 1000;

type Color = u32; // BLUE = 0, RED = 1, YELLOW = 2

fn chameneos(mut col: Color,
    mall: SyncSender<(Color, SyncSender<Option<Color>>)>) {
    // let (tx, rx) = sync_channel(0); // alternative
    loop {
        // in forest
        // let tx1 = SyncSender::clone(&tx); // alternative
        // mall.send((col, tx1)).unwrap();   // alternative
        let (tx, rx) = sync_channel(0);
        mall.send((col, tx)).unwrap();
        // waiting to meet
        match rx.recv().unwrap() {
            None => return,
            Some(other_col) =>
                if col != other_col {col = 3 - col - other_col;}
        }
    }
}
fn mall(cham: Receiver<(Color, SyncSender<Option<Color>>)>, reps: u32){
  let mut diff = 0;
  for _ in 0 .. reps {
    let (fst_col, fst_reply) = cham.recv().unwrap();
    let (snd_col, snd_reply) = cham.recv().unwrap();
    fst_reply.send(Some(snd_col)).unwrap();
    snd_reply.send(Some(fst_col)).unwrap();
    if fst_col != snd_col {diff += 1}
  }
  println!("Color changes: {}", diff);
  for _ in 0 .. CHAMS {
      let (_col, reply) = cham.recv().unwrap();
      reply.send(None).unwrap();
  }
}

fn main() {
    let (tx, rx) = sync_channel(0);
    for i in 1 .. CHAMS {
        let tx1 = SyncSender::clone(&tx);
        spawn(move || chameneos(i % 3, tx1));
    }
    spawn(move || chameneos(0, tx));
    let handle = spawn(move || mall(rx, CHAMS * MEETS / 2));
    handle.join().unwrap()
}
